source/Ship.cpp

```diff
@@ -1575,6 +1575,36 @@ void Ship::Launch(list<shared_ptr<Ship>> &ships)
 	for(Bay &bay : bays)
 		if(bay.ship && bay.ship->Commands().Has(Command::DEPLOY) && !Random::Int(40 + 20 * bay.isFighter))
 		{
+			// Determine which of the fighter's weapons we can restock.
+			set<const Outfit *> toRefill;
+			for(const auto &hit : bay.ship->Weapons())
+				if(hit.GetOutfit() && hit.GetOutfit()->Ammo() && OutfitCount(hit.GetOutfit()->Ammo()))
+					toRefill.insert(hit.GetOutfit()->Ammo());
+			// Transfer as much ammo as the fighter needs.
+			for(const Outfit *outfit : toRefill)
+			{
+				int neededAmmo = bay.ship->attributes.CanAdd(*outfit, OutfitCount(outfit));
+				if(neededAmmo)
+				{
+					AddOutfit(outfit, -neededAmmo);
+					bay.ship->AddOutfit(outfit, neededAmmo);
+					carriedMass += outfit->Mass() * neededAmmo;
+				}
+			}
```

Comment:
Does this mean that a fighter will "take" ammo that's installed in the carrier? I'm not sure that that's a desirable behavior.

Reply:
Indeed it does. The other option is to pull from cargo, but then you cannot make sure the carrier has the ammo (vs some other ship in the fleet). Pulling from cargo also means you can resupply much more easily from flotsam.

Reply:
We could augment the `Ship::Carry` routine to have the carried ship unload its ammo stores back into the carrier, as this PR does with fuel. That would maximize availability of ammo to the carrier (while fighters repair or are kept in bays), while also promoting maximal restocking of any launching fighters.

Once this restocking behavior is known to content creators, it can be pretty easy to build AI fleets that use it appropriately - stocking the AI ship with the ammo boxes, and not the launcher, is sufficient.

I admit, having a carried ship restock from the carrier could be initially confusing to players, but after they observe it happening (in either their fleet or in AI fleets), I think it will make more sense than not. This behavior has been requested before, on both Steam and here on GitHub.

https://github.com/endless-sky/endless-sky/issues/1939#issuecomment-269992593
https://github.com/endless-sky/endless-sky/issues/2336
https://steamcommunity.com/app/404410/discussions/2/492378265890026032/
https://steamcommunity.com/app/404410/discussions/0/1319962683445095725/#c1327844097131431281
(I'm sure there are other references as well)

Reply:
In current warfare, carriers tend to be focused on sending out fighting units, and resupplying them, usually meaning they have a token defensive armament built into the ship, which tends to be something like CIWS/AMT.  The Carriers in the game are right now, more like battleships, especially given the rather pathetic strength of fighters (in my opinion and playstyle).  Having them at least be able to reload their fighters/drones seems like a commonsense addition to the game; perhaps allowing fighters/drones to be re-balanced as more durable, while removing weaponry from fighter/drone carrying ships.  
This also lends itself to the idea that the player could either focus on the fighters/drones dealing them damage, or focus on the carrier itself, to prevent them from rearming.

Reply:
Personally I'd prefer fighters pull ammo from cargo, and, when assigning cargo, ships favour having ammo on a carrier that has fighters that can use it. It isn't logical having to pull missiles out of the ship's launch tube to resupply fighters, and if the player can use floatsam (assuming it even fits the fighters), then going to fish for them during the battle could be a fun playing experience.

Reply:
> Does this mean that a fighter will "take" ammo that's installed in the carrier? I'm not sure that that's a desirable behavior.

Conceptually, it's quite desirable. The entire purpose of carrier vessels is to arm, launch, and recover their carried fighters, generating what's known as a "sortie". Maximizing sortie rate by re-arming and re-fueling the fighters as quickly as possible is a key element of military aviation, particularly naval aviation. To be more realistic, Endless Sky carriers should be absolutely stuffed to the gills with ammo storage outfits (or ammunition in their cargo holds, depending on how the reload mechanic is implemented).

Reply:
A thought: Divorce carrier own ammo supply from what is intended for the fighters. Storage facilities make it possible to launch the ammo; this is not the case for the ammo that is intended for fighters. We also have to ensure that the ammo is in that ship.

The solution: Introduce outfits that convert carrier's *cargo space* without losses for bulk storage of said ammo. It can only replenish fighters, not be launched from the carrier.

How to handle the situation if both carrier and its fighters use the same ammo? When landed, always assume carrier's own weapons are loaded first, and the rest goes to the fighter ammo storage.

Reply:
Hmm, having separate "fighter ammo storage" outfits seems like it would add a lot of complication for not a lot of real gain.

Reply:
Given that this PR resolves #4112, I think I will pull the ammo-refilling changes from this PR and open those alone in a new PR.